### ISSUE TITLE
Libs install fail w/ different errors on clean mac/linux envs

### ISSUE THREAD

**Issue by agur** [chjch/pylusat#3]
- Getting failure on installation on python 3.9.13 on mac and 3.10.6 on linux (Ubuntu 22.04 LTS).  And tried on fresh virtual envs on both mac and linux.  Got different errors on either, and logged below.
- Got a fix which I can send separately as a suggestion PR.  It's a library update, w/ fixed versions (not dynamically getting latest, more stable), yet also added libs needed for running the jupyter notebooks.  
  - Also was not using the `setup.sh file`, standard attempts to use it didn't work thus resorted to directly using `requirements.txt`.
- Clean new virtualenv test command to predictably reproduce:
  - Test w/ [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/install.html):
  - `time { deactivate; rmvirtualenv lusa; mkvirtualenv lusa; python -m pip install --upgrade pip && pip install -r requirements.txt && pytest; }`
- Failure on linux:
  - ```...
            numpy/core/src/multiarray/scalartypes.c.src:3325:31: error: incompatible type for argument 1 of ‘_Py_HashDouble’
             3325 |     hashimag = _Py_HashDouble((double)
                  |                               ^~~~~~~~
                  |                               |
                  |                               double
             3326 |             (((PyC@name@ScalarObject *)obj)->obval).imag);
                  |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            In file included from /usr/include/python3.10/Python.h:77,
                             from numpy/core/src/multiarray/scalartypes.c.src:3:
            /usr/include/python3.10/pyhash.h:10:38: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘double’
               10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
                  |                                      ^~~~~~~~~~
            numpy/core/src/multiarray/scalartypes.c.src:3325:16: error: too few arguments to function ‘_Py_HashDouble’
             3325 |     hashimag = _Py_HashDouble((double)
                  |                ^~~~~~~~~~~~~~
            In file included from /usr/include/python3.10/Python.h:77,
                             from numpy/core/src/multiarray/scalartypes.c.src:3:
            /usr/include/python3.10/pyhash.h:10:23: note: declared here
               10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
                  |                       ^~~~~~~~~~~~~~
            numpy/core/src/multiarray/scalartypes.c.src: In function ‘half_arrtype_hash’:
            numpy/core/src/multiarray/scalartypes.c.src:3341:27: error: incompatible type for argument 1 of ‘_Py_HashDouble’
             3341 |     return _Py_HashDouble(npy_half_to_double(((PyHalfScalarObject *)obj)->obval));
                  |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                  |                           |
                  |                           double
            In file included from /usr/include/python3.10/Python.h:77,
                             from numpy/core/src/multiarray/scalartypes.c.src:3:
            /usr/include/python3.10/pyhash.h:10:38: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘double’
               10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
                  |                                      ^~~~~~~~~~
            numpy/core/src/multiarray/scalartypes.c.src:3341:12: error: too few arguments to function ‘_Py_HashDouble’
             3341 |     return _Py_HashDouble(npy_half_to_double(((PyHalfScalarObject *)obj)->obval));
                  |            ^~~~~~~~~~~~~~
            In file included from /usr/include/python3.10/Python.h:77,
                             from numpy/core/src/multiarray/scalartypes.c.src:3:
            /usr/include/python3.10/pyhash.h:10:23: note: declared here
               10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
                  |                       ^~~~~~~~~~~~~~
            numpy/core/src/multiarray/scalartypes.c.src: In function ‘longdouble_arrtype_hash’:
            numpy/core/src/multiarray/scalartypes.c.src:3312:1: warning: control reaches end of non-void function [-Wreturn-type]
             3312 | }
                  | ^
            numpy/core/src/multiarray/scalartypes.c.src: In function ‘float_arrtype_hash’:
            numpy/core/src/multiarray/scalartypes.c.src:3312:1: warning: control reaches end of non-void function [-Wreturn-type]
             3312 | }
                  | ^
            numpy/core/            numpy/core/src/multiarray/scalartypes.c.src: In function ‘half_arrtype_hash’:
            numpy/core/src/multiarray/scalartypes.c.src:3342:1: warning: control reaches end of non-void function [-Wreturn-type]
             3342 | }
                  | ^
            error: Command "x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -Ibuild/src.linux-x86_64-3.1/numpy/core/src/umath -Ibuild/src.linux-x86_64-3.1/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.1/numpy/core/src/common -Inumpy/core/include -Ibuild/src.linux-x86_64-3.1/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/home/vagrant/.virtualenvs/lusa/include -I/usr/include/python3.10 -Ibuild/src.linux-x86_64-3.1/numpy/core/src/common -Ibuild/src.linux-x86_64-3.1/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.1/numpy/core/src/common -Ibuild/src.linux-x86_64-3.1/numpy/core/src/npymath -c build/src.linux-x86_64-3.1/numpy/core/src/multiarray/scalartypes.c -o build/temp.linux-x86_64-cpython-310/build/src.linux-x86_64-3.1/numpy/core/src/multiarray/scalartypes.o -MMD -MF build/temp.linux-x86_64-cpython-310/build/src.linux-x86_64-3.1/numpy/core/src/multiarray/scalartypes.o.d" failed with exit status 1
            [end of output]
```
- Failure on Mac:
  - ```...
      Warning: Unused dummy argument 'itry' at (1) [-Wunused-dummy-argument]
      stat.h:8:35:
      
      Warning: Unused variable 't4' declared at (1) [-Wunused-variable]
      stat.h:8:39:
      
      Warning: Unused variable 't5' declared at (1) [-Wunused-variable]
      scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/zneupd.f:520:17:
      
        520 |             jj = workl(bounds + ncv - j)
            |                 1
      Warning: Possible change of value in conversion from COMPLEX(8) to INTEGER(4) at (1) [-Wconversion]
      stat.h:8:19:
      
      Warning: Unused variable 't0' declared at (1) [-Wunused-variable]
      stat.h:8:23:
      
      Warning: Unused variable 't1' declared at (1) [-Wunused-variable]
      stat.h:8:27:
      
      Warning: Unused variable 't2' declared at (1) [-Wunused-variable]
      stat.h:8:31:
      
      Warning: Unused variable 't3' declared at (1) [-Wunused-variable]
      stat.h:8:35:
      
      Warning: Unused variable 't4' declared at (1) [-Wunused-variable]
      stat.h:8:39:
      
      Warning: Unused variable 't5' declared at (1) [-Wunused-variable]
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/iswap.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/second_NONE.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/zmout.f
      scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/second_NONE.f:17:27:
      
         17 |       REAL               T1
            |                           1
      Warning: Unused variable 't1' declared at (1) [-Wunused-variable]
      scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/second_NONE.f:20:36:
      
         20 |       REAL               TARRAY( 2 )
            |                                    1
      Warning: Unused variable 'tarray' declared at (1) [-Wunused-variable]
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/smout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/icnteq.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/ivout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/icopy.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/ssgets.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/cvout.f
      stat.h:8:27:
      
      Warning: Unused variable 't2' declared at (1) [-Wunused-variable]
      stat.h:8:31:
      
      Warning: Unused variable 't3' declared at (1) [-Wunused-variable]
      stat.h:8:35:
      
      Warning: Unused variable 't4' declared at (1) [-Wunused-variable]
      stat.h:8:39:
      
      Warning: Unused variable 't5' declared at (1) [-Wunused-variable]
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/zvout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/svout.f
      gfortran:f77: scipy/sparse/linalg/eigen/arpack/ARPACK/UTIL/dmout.f
      gfortran:f77: /private/var/folders/1n/rm6tc95s6ng8_qwhc31w9t4h0000gn/T/pip-install-ez04p9wj/scipy_11564edbbb4848458838c09bc40640f2/scipy/_build_utils/src/wrap_dummy_g77_abi.f
      error: Command "/usr/local/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops -Iscipy/sparse/linalg/eigen/arpack/ARPACK/SRC -I/private/var/folders/1n/rm6tc95s6ng8_qwhc31w9t4h0000gn/T/pip-build-env-h37y_4mq/overlay/lib/python3.9/site-packages/numpy/core/include -c -c scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaup2.f -o build/temp.macosx-10.15-x86_64-cpython-39/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/cnaup2.o" failed with exit status 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for scipy
Failed to build scipy
ERROR: Could not build wheels for scipy, which is required to install pyproject.toml-based projects
```

**Comment by agur** [comment]
- Actually, I see that the github flow also changed from `setup.sh` to `requirements.txt` ([here](https://github.com/chjch/pylusat/commit/8fedac5aab320d08d13e4befe8d853f9f58cd446)).
- @chjch - I see no PRs here - and commits straight to the master - lmk if ok to commit to master, or would a PR be preferable (always easy to revert anyhow ...).

**Comment by chjch** [comment]
@agur Pull Request is preferable. Installation should be as simple as a `pip install pylusat` on MacOS (I'm on MacOS Monterey). pip will automatically install the dependencies such as GDAL, fiona, geopandas. 

> * Actually, I see that the github flow also changed from `setup.sh` to `requirements.txt` ([here](https://github.com/chjch/pylusat/commit/8fedac5aab320d08d13e4befe8d853f9f58cd446)).
> * @chjch - I see no PRs here - and commits straight to the master - lmk if ok to commit to master, or would a PR be preferable (always easy to revert anyhow ...).

**Comment by agur** [comment]
@chjch - `pip install pylusat` is for usage as a library.  But I was trying to setup the project locally + started running the jupyter notebooks.  Will send a PR.  I also ran the clean virtualenv on a fresh ubuntu LTS install, so it's as clean env as I could imagine it.

**Comment by agur** [comment]
Created as a PR #4 - please revise and merge, or lmk if anything else is needed.

**Comment by agur** [comment]
@chjch - still need to update this one, found that:
- Unfortunately, despite being the most stable, the fixed versions syntax (==) doesn't work across different python versions.  It fixed 3.10, but broke 3.6 for instance.  Unless we want to maintain one set of libs for each supported python then we rather revert every lib needed to >= (and keep those possible w/ ==) 
- That's why >= or even no version are so popular.  Still they force a situation w/ ever changing and unstable build.set of libs matching the latest for the oldest supported python version.
- Next I tried to start w/ == for older versions for 3.6, then move to >= for only the ones needing it.  Will follow to re-send another PR for the result.
- [act](https://github.com/nektos/act) is the best for local testing that I could find (only done on mac, not yet on linux).
